[dynamo] add common methods to DistributedVariable

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117590

This PR refactors the distributed related variables to use
DistributedVariable for common methods, so that things like
`python_type` works for all distributed variables.

Maybe we can add `as_python_constant` to the DistributedVariable too? I
didn't add in this PR but if that make sense I can update.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

wanchaol committed:
[dynamo] add common methods to DistributedVariable

This PR refactors the distributed related variables to use
DistributedVariable for common methods, so that things like
`python_type` works for all distributed variables.

Maybe we can add `as_python_constant` to the DistributedVariable too? I
didn't add in this PR but if that make sense I can update.

[ghstack-poisoned]

wconstab commented on torch/_dynamo/variables/distributed.py:
```diff
@@ -10,10 +10,18 @@
             unimplemented("torch.distributed package is not available!")
+        self.value = value
+
+    def python_type(self):
```
I think this is ok.  I would like to have a docstring on the base class Variable (where `python_type` is unimplemented) which explains under what assumptions you are allowed to implement it (or shouldn't).  Otherwise, i just look at all the other variable classes, determine this looks about the same, and ship it

cc @voznesenskym

wanchaol replied:
updated the base VariableTracker with a docstr, I don't actually know when we should not implement `python_type` so only added a comment about when we should.

wanchaol committed:
Update on "[dynamo] add common methods to DistributedVariable"


This PR refactors the distributed related variables to use
DistributedVariable for common methods, so that things like
`python_type` works for all distributed variables.

Maybe we can add `as_python_constant` to the DistributedVariable too? I
didn't add in this PR but if that make sense I can update.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

wconstab commented:
ok, will defer to @voznesenskym to stamp based on correctness of your docstring :)

voznesenskym commented on torch/_dynamo/variables/base.py:
```diff
@@ -230,6 +230,7 @@ def __repr__(self):
         return str(self)
 
     def python_type(self):
+        """Any variable that have a corresponding python type can override this method to return its python type"""
```
If you are going to touch this, go all the way, add the return `type` annotation too

```
def python_type(self):
    """
    Abstract method to be implemented by subclasses of VariableTracker.

    This method should return the type represented by the instance of the subclass. 
    The purpose is to provide a standardized way to retrieve the Python type information 
    of the variable being tracked. 

    Returns:
        type: The Python type (such as int, str, list, etc.) of the variable tracked by 
              the subclass. If the type cannot be determined or is not relevant, 
              leaving it undefined or invoking super() is always sound. 
    
    Note:
        This is an abstract method and may be overridden in subclasses.

    Example:
        class SetVariable(VariableTracker):
            def python_type(self):
                return set

    Raises:
        NotImplementedError: If the method is not implemented in a subclass.
    """
    ```

wanchaol replied:
Thanks for the detailed docstr!

voznesenskym commented on torch/_dynamo/variables/distributed.py:
```diff
@@ -10,10 +10,19 @@
 
 class DistributedVariable(VariableTracker):
-    def __init__(self, **kwargs):
+    """
+    The base distributed variable that checks for the distributed existance
+    and hold the value for a distributed object.
+    """
```
Not sure this comment says much.

wanchaol replied:
Updated with more content :)

voznesenskym reviewed (approved):
Good PR, just nits

wanchaol committed:
Update on "[dynamo] add common methods to DistributedVariable"


This PR refactors the distributed related variables to use
DistributedVariable for common methods, so that things like
`python_type` works for all distributed variables.

Maybe we can add `as_python_constant` to the DistributedVariable too? I
didn't add in this PR but if that make sense I can update.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge

wanchaol commented:
@pytorchbot merge -i
